File: Server/Src/Zones/ZoneHandler.cs
using System.Text.Json;

public class ZoneHandler
{
    private static ZoneHandler instance;
    private readonly ZoneManager zoneManager = ZoneManager.GetInstance();
    private readonly ZonesDataManager zonesDataManager = ZonesDataManager.GetInstance();
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNameCaseInsensitive = true
    };
    
    private ZoneHandler()
    {
    }
    public static ZoneHandler GetInstance()
    {
        if (instance == null)
        {
            instance = new ZoneHandler();
        }
        return instance;
    }

    public void HandleAddZone(JsonElement data, ModeEnum clientMode)
    {
        try
        {
            Zone zone = JsonSerializer.Deserialize<Zone>(data);

            Guid uuid = Guid.NewGuid();
            string uuidString = uuid.ToString();
            zone.zoneId = uuidString;


            // add in file
            zonesDataManager.AddAndSaveZone(zone);

            // add in a map 
            bool isAdded = zoneManager.TryAddZone(zone);
            if (isAdded)
            {
                System.Console.WriteLine("{0} ({1}) - Added zone successfully.", zone.zoneId, zone.zoneName);
                SendAddZone(zone, clientMode);
            }
            else
            {
                System.Console.WriteLine("{0} ({1}) - Failed to add zone.", zone.zoneId, zone.zoneName);
                SendZoneError($"{zone.zoneId} ({zone.zoneName}) - Failed to add zone.", clientMode);
            }
        }
        catch (Exception ex)
        {
            System.Console.WriteLine("Error in HandleAddZone: " + ex.Message);
            System.Console.WriteLine("AAAAA");
        }
    }

    public void HandleRemoveZone(JsonElement data, ModeEnum clientMode)
    {
        try
        {
            Zone zone = data.Deserialize<Zone>();
            string zoneId = zone.zoneId;

            bool isRemoved = zonesDataManager.RemoveAndSaveZone(zoneId);
            if (isRemoved)
            {
                isRemoved = zoneManager.TryRemoveZone(zoneId); 
                if (isRemoved)
                {
                    System.Console.WriteLine("{0} ({1}) - Removed zone successfully.", zoneId, zone.zoneName);
                    SendRemoveDangerZone(zone, clientMode);
                }
                else
                {
                    System.Console.WriteLine("{0} ({1}) - Failed to remove zone.", zoneId, zone.zoneName);
                    SendZoneError($"{zoneId} ({zone.zoneName}) - Failed to remove zone.", clientMode);
                }
            }
            else
            {
                System.Console.WriteLine("{0} ({1}) - Failed to remove zone.", zoneId, zone.zoneName);
                SendZoneError($"{zoneId} ({zone.zoneName}) - Failed to remove zone from file.", clientMode);
            }
        }
        catch (Exception ex)
        {
            System.Console.WriteLine("Error in HandleRemoveDangerZone: " + ex.Message);
        }
    }

    public void HandleEditZone(JsonElement data, ModeEnum clientMode)
    {
        try
        {
            Zone zone = data.Deserialize<Zone>();
            string zoneId = zone.zoneId;

            bool isEdited = zonesDataManager.EditAndSaveZone(zoneId, zone); 
            if (isEdited)
            {
                isEdited = zoneManager.TryEditZone(zoneId, zone); 
                if (isEdited)
                {
                    System.Console.WriteLine("{0} ({1}) - Edited zone successfully.", zoneId, zone.zoneName);
                    SendEditDangerZone(zone, clientMode);
                }
                else
                {
                    System.Console.WriteLine("{0} ({1}) - Failed to edit zone.", zoneId, zone.zoneName);
                    SendZoneError($"{zoneId} ({zone.zoneName}) - Failed to edit zone.", clientMode);
                }
            }
            else
            {
                System.Console.WriteLine("{0} ({1}) - Failed to edit zone.", zoneId, zone.zoneName);
                SendZoneError($"{zoneId} ({zone.zoneName}) - Failed to edit zone from file.", clientMode);
            }
        }
        catch (Exception ex)
        {
            System.Console.WriteLine("Error in HandleEditDanger: " + ex.Message);
        }
    }

    public void SendAddZone(Zone zone, ModeEnum clientMode)
    {
        string zoneData =  WebSocketServer.prepareMessageToClient(S2CMessageType.AddZone, zone, clientMode);
        WebSocketServer.SendMsgToClients(zoneData, clientMode);
    }
    public void SendRemoveDangerZone(Zone zone, ModeEnum clientMode)
    {
        string zoneData =  WebSocketServer.prepareMessageToClient(S2CMessageType.RemoveZone, zone, clientMode);
        WebSocketServer.SendMsgToClients(zoneData, clientMode);
    }
    public void SendEditDangerZone(Zone zone, ModeEnum clientMode)
    {
        string zoneData =  WebSocketServer.prepareMessageToClient(S2CMessageType.EditZone, zone, clientMode);
        WebSocketServer.SendMsgToClients(zoneData, clientMode);
    }
    public void SendZoneError(string errorMsg, ModeEnum clientMode)
    {
        ZoneError zoneError = new ZoneError()
        {
            errorMsg = errorMsg
        };
        string zoneErrorData = WebSocketServer.prepareMessageToClient(S2CMessageType.ZoneError, zoneError, clientMode);
        WebSocketServer.SendMsgToClients(zoneErrorData, clientMode);
    }
}

File: ReactClient/src/Handlers/ZoneHandler.ts
import { ZoneEntityManager } from "../ZonesPanel/ZoneEntityManager";
import * as Cesium from "cesium";
import type { DangerZone, ZoneError as ZoneError, Zone } from "../Messages/AllTypes";
import { toast } from "react-toastify";
import { ZoneManager } from "../Managers/ZoneManager";

export class ZoneHandler{
    private static instance: ZoneHandler | null = null;
    private zoneManager: ZoneManager
    private zoneEntityManager: ZoneEntityManager
    private constructor(viewer: Cesium.Viewer) {
        this.zoneManager = ZoneManager.getInstance();
        this.zoneEntityManager = ZoneEntityManager.GetInstance(viewer);
    }

    public static getInstance(viewer: Cesium.Viewer): ZoneHandler {
        if (this.instance === null) {
            this.instance = new ZoneHandler(viewer);
        }
        return this.instance;
    }
    
    
    HandleAddZone(data: any){
        try {
            const zone = data as Zone;
            this.AddZone(zone)
        } catch (err) {
            console.log("data could not be parsed toZone");
        }
    }
    AddZone(zone: Zone){
        const isAdded = this.zoneManager.tryAddZone(zone);
        if(isAdded){
            this.zoneEntityManager.tryAddZone(zone);
            console.log(` zone ${zone.zoneId} added successfully.`);
        }
        else
            console.log("error in HandleAddZone.  zone adding failed")
    }

    HandleRemoveZone(data: any){
        try {
            const zone = data as Zone;
            this.RemoveZone(zone)
        } catch (err) {
            console.log("data could not be parsed to Zone");
        }
    }

    RemoveZone(zone: Zone){
        const isRemoved = this.zoneManager.tryRemoveZone(zone.zoneId);
        if(isRemoved){
            this.zoneEntityManager.removeZone(zone.zoneId);
            console.log(`zone ${zone.zoneId} removed successfully.`);
        }
        else
            console.log("error in HandleRemoveZone. zone removing failed")
    }

    HandleEditZone(data: any){
        try {
            const zone = data as Zone;
            this.EditZone(zone);
            
        } catch (err) {
            console.log("data could not be parsed to Zone");
        }
    }

    EditZone(zone: Zone){
        const isEdited = this.zoneManager.tryEditZone(zone);
        if(isEdited){
            this.zoneEntityManager.editZone(zone);
            console.log(`danger zone ${zone.zoneId} edited successfully.`);
        }
        else
            console.log("error in HandleEditDangerZone. danger zone editing failed")
    }

    HandleZoneError(data: any){
        try {
            const zoneError = data as ZoneError;
            const errorMsg = zoneError.errorMsg;
            console.log("zone error from server: " + errorMsg);
            toast.error(errorMsg);
        } catch (err) {
            console.log("data could not be parsed to ZoneError");
        }
    }
}